### DISCUSSION TITLE
I need to run sha256sum on the device, from adb shell, how can I accomplish this task ?

### DISCUSSION THREAD

**Discussion by shodanx2** [termux/termux-app#3290] [Q&A] [answered]
Hi,

My initial problem, I need to copy some files over and I have to make sure they are intact.

So I would like to run  sha256sum myfile > myfile.sha256
Then using adb pull, download both myfile and myfile.sha256
And finally, on my local host, compare sha256sum myfile with myfile.sha256

So after a little bit of search, I find termux or termux-nix are the way to go

BTW this is an unrootable Moto Z3 and Moto Z4 phones

I have sideloaded termux, on the phone using scrcpy I can run termux on my phone, I can run sha256sum just fine.

But this is a fully automated setup I'm making, I'm going to make a bash script that does all this from my local computer remotely to the phone (using adb tcpip over wifi)

So using adb -s myphone.lan shell  and I try to start termux on there

According to this issue https://github.com/termux/termux-app/issues/77

The shell executable should be at 
/data/data/com.termux/files/usr/bin/bash
However

```
127|foles:/ $ /data/data/com.termux/files/usr/bin/bash
/system/bin/sh: /data/data/com.termux/files/usr/bin/bash: inaccessible or not found
127|foles:/ $ /data/data/com.termux/files/usr/bin/zsh
/system/bin/sh: /data/data/com.termux/files/usr/bin/zsh: inaccessible or not found
```

And I can't seem to be able to see any of those files

```
127|foles:/ $ ls -lsh /data/data/com.termux/files/usr/bin/
ls: /data/data/com.termux/files/usr/bin/: Permission denied
1|foles:/ $ ls -lsh /data/data/com.termux/files/usr/
ls: /data/data/com.termux/files/usr/: Permission denied
1|foles:/ $ ls -lsh /data/data/com.termux/files/
ls: /data/data/com.termux/files/: Permission denied
1|foles:/ $ ls -lsh /data/data/com.termux/
ls: /data/data/com.termux/: Permission denied
1|foles:/ $ ls -lsh /data/data/com.termux1
ls: /data/data/com.termux1: No such file or directory
```

The above issues says "connect using USB"  but that's really about tunnelling ports over adb to connect to sshd running on the device.
My device does not currently run a sshd server. 
Maybe I should ? But it would be nice if adb would "just work" because now to sha256sum, I need to start sshd, connect to sshd then run sha256sum. I imagine the sshd will die when device goes to sleep and certainly everytime I reboot the device ? 

It feels like an extra layer of flake, and adb is already super flakey, stops working everytime the device reboot, I need to connect it to usb and run adb -s SERIALNUMBER tcpip 5555 every, single, time !!!!

I'm really hoping with termux, I'll be able to run a bash script that starts adb over tcpip on the phone every time it reboots ?

So what solution should I pursue next ?

Thanks !

**Comment by agnostic-apollo** [comment, marked as answer]
I assume you are pulling files for public storage/sd card. For that on android `>= 8` you can use system provided `sha256sum` binary in `/system/bin` with `adb shell sha256sum`.

https://android.googlesource.com/platform/system/core/+/master/shell_and_utilities/README.md

>I'm really hoping with termux, I'll be able to run a bash script that starts adb over tcpip on the phone every time it reboots ?

That's not possible due to android security restrictions without root.

You can start a termux `sshd -D` server in foreground at boot with termux-boot so it is less likely to be killed and it won't require adb, just both devices being connected to same wifi.

If you want to use termux over adb, you can install termux debuggable builds from github and use `run-as`.

https://github.com/termux/termux-app/discussions/3249

https://github.com/termux/termux-app#Installation

**Comment by shodanx2** [comment]
Ok I tried install sshd, this does not work, I have no internet connection, this phone is purposedly prevented from talking to the internet so that it will not download Verizon killswitch and kill itself.


```
~ $ sshd
The program sshd is not installed. Install it by executing:
 pkg install openssh
~ $ pkg install openssh
Testing the available mirrors:
[*] https://packages-cf.termux.org/apt/termux-main: bad
[*] https://deb.kcubeterm.me/termux-main: bad
[*] https://termux.mentality.rip/termux-main: bad
[*] https://grimler.se/termux-packages-24: bad
[*] https://termux.librehat.com/apt/termux-main: bad
Using fallback mirror: https://packages-cf.termux.org/apt/termux-main
0% [Connecting to packages-cf.termux.org]^C
```

I tried searching a method of offline installation, where I would download the sshd package from my local computer, then push it to the right place and have termux install it from there

However my search of how to do this have not yielded a solution yet

This was apparently issue #1 but was evidently never fixed

https://github.com/termux/termux-app/issues/1

Ok I found another potential solution on reddit

https://old.reddit.com/r/termux/comments/ge3ckm/offline_package_installation/

Download the entire repo using the script

https://github.com/termux/termux-packages/blob/master/scripts/buildorder.py

```
root@nmap:~/termuxdownload# wget https://raw.githubusercontent.com/termux/termux                                                                                                                                                                       -packages/master/scripts/buildorder.py
root@nmap:~/termuxdownload# chmod +x buildorder.py
root@nmap:~/termuxdownload# ./buildorder.py
Traceback (most recent call last):
  File "/root/termuxdownload/./buildorder.py", line 301, in <module>
    main()
  File "/root/termuxdownload/./buildorder.py", line 290, in main
    pkgs_map = read_packages_from_directories(packages_directories, fast_build_mode, full_buildorder)
  File "/root/termuxdownload/./buildorder.py", line 155, in read_packages_from_directories
    with open ('repo.json') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'repo.json'
```

Ok, looks like I also need the repo file

Ah, there is it 

https://github.com/termux/termux-packages/blob/master/repo.json

```
root@nmap:~/termuxdownload# wget https://raw.githubusercontent.com/termux/termux-packages/master/repo.json
root@nmap:~/termuxdownload# ./buildorder.py
Traceback (most recent call last):
  File "/root/termuxdownload/./buildorder.py", line 301, in <module>
    main()
  File "/root/termuxdownload/./buildorder.py", line 290, in main
    pkgs_map = read_packages_from_directories(packages_directories, fast_build_mode, full_buildorder)
  File "/root/termuxdownload/./buildorder.py", line 160, in read_packages_from_directories
    for pkgdir_name in sorted(os.listdir(package_dir)):
FileNotFoundError: [Errno 2] No such file or directory: 'packages'
```

Ok, going to need one more file !
Oh..
![image](https://user-images.githubusercontent.com/10621885/223964826-2e2600ac-f036-4434-8cf0-7239dff2591e.png)
Ok, maybe it's easier to just download the entire thing

```
root@nmap:~/termuxdownload# git clone https://github.com/termux/termux-packages
root@nmap:~/termuxdownload# ls
buildorder.py  repo.json  termux-packages
root@nmap:~/termuxdownload# cd termux-packages/
root@nmap:~/termuxdownload/termux-packages# ls
CODEOWNERS  CONTRIBUTING.md  LICENSE.md  README.md  build-all.sh  build-package.sh  clean.sh  disabled-packages  ndk-patches  packages  repo.json  root-packages  sample  scripts  x11-packages
root@nmap:~/termuxdownload/termux-packages# ./scripts/buildorder.py
```

This produced a very long output, but no downloads
Maybe this should give a better output, just openssh
```

root@nmap:~/termuxdownload/termux-packages# ./scripts/buildorder.py packages/openssh
ERROR: Package scrcpy depends on non-existing package "sdl2"
```

Weird, what does scrcpy have to do with any of this ?

Maybe I can find a clue in the package folder

https://github.com/termux/termux-packages/tree/master/packages/openssh

Ah, there's an URL !

![image](https://user-images.githubusercontent.com/10621885/223966588-852223b7-fc7a-456c-b3bc-c18697d11d4a.png)

Ok, I need https://fastly.cdn.openbsd.org/pub/OpenBSD/OpenSSH/portable/openssh-9.2p1.tar.gz

As well as dependencies TERMUX_PKG_DEPENDS="krb5, ldns, libandroid-support, libedit, openssh-sftp-server, openssl, termux-auth, zlib"


On a windows box for a moment

```
E:\camdownload>curl https://fastly.cdn.openbsd.org/pub/OpenBSD/OpenSSH/portable/openssh-9.2p1.tar.gz -o openssh-9.2p1.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1808k  100 1808k    0     0  5669k      0 --:--:-- --:--:-- --:--:-- 5688k

E:\camdownload>adb -s cam1.lan push openssh-9.2p1.tar.gz /mnt/user/0/primary/
openssh-9.2p1.tar.gz: 1 file pushed, 0 skipped. 309.5 MB/s (1852380 bytes in 0.006s)
```

Ok, first package pushed !


ok some more dependencies

```
TERMUX_PKG_VERSION=1.20.1
TERMUX_PKG_SRCURL=https://fossies.org/linux/misc/krb5-$TERMUX_PKG_VERSION.tar.gz
TERMUX_PKG_DEPENDS="libandroid-support, libandroid-glob, libresolv-wrapper, readline, openssl, libdb"

TERMUX_PKG_VERSION=1.8.3
TERMUX_PKG_SRCURL=https://www.nlnetlabs.nl/downloads/ldns/ldns-${TERMUX_PKG_VERSION}.tar.gz
TERMUX_PKG_DEPENDS="openssl, resolv-conf"

TERMUX_PKG_VERSION=(28
		    3)
TERMUX_PKG_SRCURL=(https://github.com/termux/libandroid-support/archive/v${TERMUX_PKG_VERSION[0]}.tar.gz
		   https://github.com/termux/wcwidth/archive/v${TERMUX_PKG_VERSION[1]}.tar.gz)

TERMUX_PKG_VERSION=20221030-3.1
TERMUX_PKG_SRCURL=https://thrysoee.dk/editline/libedit-${TERMUX_PKG_VERSION}.tar.gz
TERMUX_PKG_DEPENDS="libandroid-support, ncurses"
```

no build.sh for openssh-sftp-server

```
_VERSION=3.0.8
TERMUX_PKG_VERSION=1:${_VERSION}
TERMUX_PKG_SRCURL=https://www.openssl.org/source/openssl-${_VERSION/\~/-}.tar.gz
TERMUX_PKG_DEPENDS="ca-certificates, zlib"

TERMUX_PKG_VERSION=1.4
TERMUX_PKG_SRCURL=https://github.com/termux/termux-auth/archive/v${TERMUX_PKG_VERSION}.tar.gz
TERMUX_PKG_DEPENDS="openssl"

TERMUX_PKG_VERSION=1.2.13
TERMUX_PKG_SRCURL=https://www.zlib.net/zlib-$TERMUX_PKG_VERSION.tar.xz
```

no download URL for libandroid-glob

```
TERMUX_PKG_VERSION=1.1.7
TERMUX_PKG_SRCURL=https://ftp.samba.org/pub/cwrap/resolv_wrapper-${TERMUX_PKG_VERSION}.tar.gz
TERMUX_PKG_DEPENDS="resolv-conf"

_MAIN_VERSION=8.2
_PATCH_VERSION=1
TERMUX_PKG_VERSION=$_MAIN_VERSION.$_PATCH_VERSION
TERMUX_PKG_SRCURL=https://mirrors.kernel.org/gnu/readline/readline-${_MAIN_VERSION}.tar.gz

TERMUX_PKG_VERSION=18.1.40
TERMUX_PKG_SRCURL=https://fossies.org/linux/misc/db-${TERMUX_PKG_VERSION}.tar.gz
TERMUX_PKG_DEPENDS="libc++"
```

No download URL for resolv-conf

```
TERMUX_PKG_VERSION=(6.4
		    9.30
		    15
		    0.26.5
		    0.9.0)
TERMUX_PKG_SRCURL=(https://ftp.gnu.org/gnu/ncurses/ncurses-${TERMUX_PKG_VERSION[0]}.tar.gz
		   https://fossies.org/linux/misc/rxvt-unicode-${TERMUX_PKG_VERSION[1]}.tar.bz2
		   https://github.com/thestinger/termite/archive/v${TERMUX_PKG_VERSION[2]}.tar.gz
		   https://github.com/kovidgoyal/kitty/archive/v${TERMUX_PKG_VERSION[3]}.tar.gz
		   https://github.com/alacritty/alacritty/archive/refs/tags/v${TERMUX_PKG_VERSION[4]}.tar.gz)

_DATE=2023-01-10
TERMUX_PKG_VERSION=1:${_DATE//-/.}
TERMUX_PKG_SRCURL=https://curl.se/ca/cacert-${_DATE}.pem

TERMUX_PKG_VERSION=25c
TERMUX_PKG_SRCURL=https://dl.google.com/android/repository/android-ndk-r${TERMUX_PKG_VERSION}-linux.zip
```

Ok That should be all of it for openssh server 

```
curl -O https://fossies.org/linux/misc/krb5-1.20.1.tar.gz  
curl -O https://www.nlnetlabs.nl/downloads/ldns/ldns-1.8.3.tar.gz
curl -O https://github.com/termux/libandroid-support/archive/v28.tar.gz
curl -O https://thrysoee.dk/editline/libedit-20221030-3.1.tar.gz
curl -O https://www.openssl.org/source/openssl-3.0.8.tar.gz
curl -O https://github.com/termux/termux-auth/archive/v1.4.tar.gz
curl -O https://www.zlib.net/zlib-1.2.13.tar.xz
curl -O https://ftp.samba.org/pub/cwrap/resolv_wrapper-1.1.7.tar.gz
curl -O https://mirrors.kernel.org/gnu/readline/readline-8.2.1.tar.gz
curl -O https://fossies.org/linux/misc/db-18.1.40.tar.gz
curl -O https://ftp.gnu.org/gnu/ncurses/ncurses-6.4.tar.gz
curl -O https://curl.se/ca/cacert-2023-01-10.pem
curl -O https://dl.google.com/android/repository/android-ndk-r25c-linux.zip
```

Ok, results

```
2023-03-09  04:02        30,763,705 db-18.1.40.tar.gz
2023-03-09  04:02         8,661,660 krb5-1.20.1.tar.gz
2023-03-09  04:02         1,299,856 ldns-1.8.3.tar.gz
2023-03-09  04:02           533,261 libedit-20221030-3.1.tar.gz
2023-03-09  04:02         3,612,591 ncurses-6.4.tar.gz
2023-03-09  03:40         1,852,380 openssh-9.2p1.tar.gz
2023-03-09  04:02        15,151,328 openssl-3.0.8.tar.gz
2023-03-09  04:02               369 readline-8.2.1.tar.gz
2023-03-09  04:02            45,148 resolv_wrapper-1.1.7.tar.gz
2023-03-09  04:02                 0 v1.4.tar.gz
2023-03-09  04:02                 0 v28.tar.gz
2023-03-09  04:02         1,296,496 zlib-1.2.13.tar.xz
```

Ok, these are wrong

https://mirrors.kernel.org/gnu/readline/readline-8.2.1.tar.gz (file doesn't exist)
https://github.com/termux/libandroid-support/archive/v28.tar.gz (folder missing)
https://github.com/termux/termux-auth/archive/v1.4.tar.gz (folder missing)

Ok got readline 8.2 instead
https://mirrors.kernel.org/gnu/readline/readline-8.2.tar.gz


So, I'm still missing 

```
libandroid-support
termux-auth
resolv-conf
libandroid-glob
openssh-sftp-server
```

Ok pushed the files I've for now, maybe it's enough

```
E:\camdownload>adb -s cam1.lan push files /mnt/user/0/primary/files/
files\: 12 files pushed, 0 skipped. 27.3 MB/s (597595153 bytes in 20.905s)
```

```
E:\camdownload>adb -s cam1.lan shell ls /mnt/user/0/primary/files/files
android-ndk-r25c-linux.zip
cacert-2023-01-10.pem
db-18.1.40.tar.gz
krb5-1.20.1.tar.gz
ldns-1.8.3.tar.gz
libedit-20221030-3.1.tar.gz
ncurses-6.4.tar.gz
openssh-9.2p1.tar.gz
openssl-3.0.8.tar.gz
readline-8.2.tar.gz
resolv_wrapper-1.1.7.tar.gz
zlib-1.2.13.tar.xz
```

Ok, now how do I make pkg use my files folder as a download cache, instead of just download them files !?

Ok, this page explains the package manager

https://wiki.termux.com/wiki/Package_Management

The cache is mentionned 

![image](https://user-images.githubusercontent.com/10621885/223975473-755382df-89eb-47f5-89a8-b27117bc8ec5.png)


Just need to make it use my folder instead !

Oh my, I just saw the time, I will have to retry tomorrow night

**Reply by agnostic-apollo** [reply]
Install with `dpkg -i /path/to/deb`. You will have to install all the dependencies in the `control` file manually.

https://packages.termux.dev/

https://packages.termux.dev/apt/termux-main/pool/main/o/openssh/

**Reply by twaik** [reply]
> dpkg -i /path/to/deb

It is also possible to use `apt install /path/to/deb`, it will also handle all dependencies.

**Comment by shodanx2** [comment]
@agnostic-apollo 

Thank you,
For sha256sum, I was trying sha256 short and it wasn't working, you are right it is sha256sum and it works !

E:\camdownload>adb -s cam1.lan shell "echo test | sha256sum "
f2ca1bb6c7e907d06dafe4687e579fce76b37e4e93b7605022da52e6ccc26fd2  -

That solves that !

But I'm sure I will need termux for other things so I will continue trying to make this work tomorrow !



    I'm really hoping with termux, I'll be able to run a bash script that starts adb over tcpip on the phone every time it reboots ?

| That's not possible due to android security restrictions without root.

If I somehow get root, what command would need to run from a boot script to make this happen on every boot ?

The obsolete OS I run, Android 9, has [quite a few security vulnerability](https://vulmon.com/searchpage?q=google+android+9.0) which I hope I can manage to get one to give me a root shell someday

--

Is there any way I can set the pkg download cache folder to my /mnt/user/0/primary/files/ folder  ?

and lastly

What would be the procedure for doing the equivalent of 

pkg upgrade

manually ?

thanks !

**Comment by agnostic-apollo** [comment]
Enable wifi mode with `setprop service.adb.tcp.port "5555" && stop adbd && start adbd` and enable usb mode with  `setprop service.adb.tcp.port "-1" && stop adbd && start adbd`

>What would be the procedure for doing the equivalent of pkg upgrade

Checking repo for updates and manually installing updates with `dpkg`.

>Is there any way I can set the pkg download cache folder to my /mnt/user/0/primary/files/ folder ?

You may instead want to consider setting up termux in a device that has internet and then backing it up and restoring it on a device that doesn't have access.

https://wiki.termux.com/wiki/Backing_up_Termux

**Reply by agnostic-apollo** [reply]
@twaik tsk tsk tsk...
